From 6df28e615c57cf43af9f9183fa8646c941ae7bd3 Mon Sep 17 00:00:00 2001
From: Billy Olsen <billy.olsen@canonical.com>
Date: Mon, 14 Aug 2023 16:17:37 -0700
Subject: [PATCH] Update mysql schema for modern mysql

The groups keyword is reserved in mysql but can still have tables
using this name if they are quoted with backticks (`). Additionally,
default values cannot be specified for TEXT types unless wrapped in
parentheses.

Signed-off-by: Billy Olsen <billy.olsen@canonical.com>
---
 mysql.go | 10 +++++-----
 1 file changed, 5 insertions(+), 5 deletions(-)

diff --git a/v2/pkg/plugins/glauth-mysql/mysql.go b/v2/pkg/plugins/glauth-mysql/mysql.go
index 914889e..db2fe6b 100644
--- a/v2/pkg/plugins/glauth-mysql/mysql.go
+++ b/v2/pkg/plugins/glauth-mysql/mysql.go
@@ -44,15 +44,15 @@ CREATE TABLE IF NOT EXISTS users (
 	passbcrypt VARCHAR(64) DEFAULT '',
 	otpsecret VARCHAR(64) DEFAULT '',
 	yubikey VARCHAR(128) DEFAULT '',
-	sshkeys TEXT DEFAULT '',
-	custattr TEXT DEFAULT '{}')
+	sshkeys TEXT DEFAULT (''),
+	custattr TEXT DEFAULT ('{}'))
 `)
 	statement.Exec()
 	statement, _ = db.Prepare("CREATE UNIQUE INDEX idx_user_name on users(name)")
 	statement.Exec()
-	statement, _ = db.Prepare("CREATE TABLE IF NOT EXISTS groups (id INTEGER AUTO_INCREMENT PRIMARY KEY, name VARCHAR(64) NOT NULL, gidnumber INTEGER NOT NULL)")
+	statement, _ = db.Prepare("CREATE TABLE IF NOT EXISTS `groups` (id INTEGER AUTO_INCREMENT PRIMARY KEY, name VARCHAR(64) NOT NULL, gidnumber INTEGER NOT NULL)")
 	statement.Exec()
-	statement, _ = db.Prepare("CREATE UNIQUE INDEX idx_group_name on groups(name)")
+	statement, _ = db.Prepare("CREATE UNIQUE INDEX idx_group_name on `groups`(name)")
 	statement.Exec()
 	statement, _ = db.Prepare("CREATE TABLE IF NOT EXISTS includegroups (id INTEGER AUTO_INCREMENT PRIMARY KEY, parentgroupid INTEGER NOT NULL, includegroupid INTEGER NOT NULL)")
 	statement.Exec()
@@ -63,7 +63,7 @@ CREATE TABLE IF NOT EXISTS users (
 // Migrate schema if necessary
 func (b MysqlBackend) MigrateSchema(db *sql.DB, checker func(*sql.DB, string) bool) {
 	if !checker(db, "sshkeys") {
-		statement, _ := db.Prepare("ALTER TABLE users ADD COLUMN sshkeys TEXT DEFAULT ''")
+		statement, _ := db.Prepare("ALTER TABLE users ADD COLUMN sshkeys TEXT DEFAULT ('')")
 		statement.Exec()
 	}
 }
-- 
2.39.2

